Make compare previous better + new legend

alexkim205 opened the pull request:
## Changes

The new legend is behind a FF called `insight-legends`

- [x] Update insight canvas header styles to match https://github.com/PostHog/posthog/issues/4705#issuecomment-895611473
- [x] Add current/previous tags in insight table
- [x] Make previous colors lighter corresponding current colors
- [x] Don't expose compare to if date filter is "All time"
- [x] Update date filter to show formatted dates (only changed for Insight Date Filter to reduce blast radius)
- [x] Add legend to trends + stickiness insights (insights with compare to functionality)
- [x] Fix various compare bugs
- [x] Add `compare_label` parameter to trend result
- [x] Add time freezing dev package
- [x] Show new legend by default with `legend_hidden` filter param
- [x] Fix breakdown visibilities in entity labels in various places: insights table, funnels step table, filters, etc.

Note: I'll tackle tooltips in another PR.

## Before

![screencapture-app-posthog-insights-FFuRBl7B-edit-2021-12-06-16_14_04](https://user-images.githubusercontent.com/13460330/144942845-bfe186bc-684a-429d-aded-9842bf01356c.png)

## After - Can you spot the differences?

With compare to previous

![screencapture-localhost-8000-insights-aJp403j8-edit-2021-12-08-11_29_13](https://user-images.githubusercontent.com/13460330/145271801-47ea0add-217b-46bf-b0fe-db9f1ffa2283.png)


With breakdown

![screencapture-localhost-8000-insights-aJp403j8-edit-2021-12-08-11_29_56](https://user-images.githubusercontent.com/13460330/145271884-3caec08a-8996-478c-b625-74665562ce59.png)



Insight types this affects:
- [x] All stickiness graph types
- [x] All trends graph types
- [x] Ignore sessions (on deprecation path)

## How did you test this code?

Added logic tests + smoke testing of new ui components.

Works with breakdowns, multi breakdowns, $current_urls, etc.

There is a bug with today's code where breakdowns + compare doesn't return data from the previous time range (see bug in action [here](https://app.posthog.com/insights/FFuRBl7B/edit?insight=TRENDS&actions=%5B%5D&events=%5B%7B%22id%22%3A%22%24pageview%22%2C%22name%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A0%7D%2C%7B%22id%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A1%2C%22name%22%3A%22%24pageview%22%7D%2C%7B%22id%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A2%2C%22name%22%3A%22%24pageview%22%7D%5D&display=ActionsLineGraph&interval=day&properties=%5B%5D&funnel_to_step=1&funnel_viz_type=steps&exclusions=%5B%5D&funnel_from_step=0&compare=true&breakdown=%24current_url&breakdown_type=event)). This PR doesn't fix this, but theoretically should work once the backend bug is fixed (tested with dummy data). @EDsCODE  do you know if this is being tracked anywhere?

alexkim205 committed:
copy

alexkim205 committed:
reorganize insight headers

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-tooltip-legend

alexkim205 committed:
wip

alexkim205 committed:
add compare_label parameter to trends and stickiness insights

alexkim205 committed:
fix labels

alexkim205 committed:
remove console

paolodamico reviewed (dismissed):
Initial pass (UI/UX only, haven't looked much at the code), it's looking pretty great!

1. I'm not sure if the new banner fits with the current design. I would suggest either adding a light gray background or if we're going to add a background to the main graph (per the Figma design). Will defer to @clarkus here.
    <img width="1183" alt="" src="https://user-images.githubusercontent.com/5864173/144773580-d0e9cd99-0f47-4652-adf0-e51accd5ffff.png">
2. I don't think this is good UX, because it's hard to parse. One approach that comes to mind is changing the header name to show both relevant dates. An alternative is just say "Day n" in the header and show the date in the cell.
    <img width="1172" alt="" src="https://user-images.githubusercontent.com/5864173/144773873-c73cb184-7741-4450-9912-5def34d03e2c.png">
3. Unsure if it's part of the scope of this PR, but tooltips are still super confusing.
    <img width="315" alt="" src="https://user-images.githubusercontent.com/5864173/144774222-1ce408f0-cfff-4757-ba96-69e07dc4f2ff.png">
4. I can still select the "compare" option even if all time is set (which btw made my browser crash, so good call here).
    <img width="697" alt="" src="https://user-images.githubusercontent.com/5864173/144774492-1989cb9c-3527-417e-974e-a1ba0b44790f.png">
5. Re date formats, not sure if we should zero-pad numbers, feels pretty common outside the US. Thoughts @clarkus ?

The new legend is amazing! Love it, some thoughts.
6. I think it might be worth vertically centering the legend card if it's small.
    <img width="313" alt="" src="https://user-images.githubusercontent.com/5864173/144775465-5256a5e8-cc2d-48c6-85b8-7610972712f4.png">
7. The behavior of opening the legend feels strange. Not sure if it's a matter of adding an animation. Any ideas @clarkus ? Also, thinking it might be worth having it open by default (as an experiment).
8. I found this bug when I created the chart in the screenshot where the legend shows empty.
    <img width="1196" alt="" src="https://user-images.githubusercontent.com/5864173/144775557-9f8fc14f-221c-45f5-8956-830c6044caf6.png">
9. Think the legend should have a max width and then just allow scrolling.
    ![image](https://user-images.githubusercontent.com/5864173/144775663-67734793-dfaa-459f-b907-cbdffaa9846d.gif)

alexkim205 committed:
fix broken funnel breakdown by multiple property

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-tooltip-legend

alexkim205 committed:
Merge branch 'fix/breakdowns-not-working-funnel-simple-mode' into feat/new-tooltip-legend

alexkim205 commented:
> 1. I'm not sure if the new banner fits with the current design. I would suggest either adding a light gray background or if we're going to add a background to the main graph (per the Figma design). Will defer to @clarkus here.

I've added a light grey bg to the insight graph container and standardized the bg colors we use across insights to bring it up to speed with our Figma mockups. 

> 2. I don't think this is good UX, because it's hard to parse. One approach that comes to mind is changing the header name to show both relevant dates. An alternative is just say "Day n" in the header and show the date in the cell.
>    <img alt="" width="1172" src="https://user-images.githubusercontent.com/5864173/144773873-c73cb184-7741-4450-9912-5def34d03e2c.png">

> 3. Unsure if it's part of the scope of this PR, but tooltips are still super confusing.

Yup, i'd prefer to chunk this out into it's own PR to avoid making this bigger. I'll tag you when that's ready!

> 4. I can still select the "compare" option even if all time is set (which btw made my browser crash, so good call here).

Fixed and wrote tests for it

> 5. Re date formats, not sure if we should zero-pad numbers, feels pretty common outside the US. Thoughts @clarkus ?

my 2c, zero-padding here will make the width ever so slightly wider than it has to be. We're already working with limited real estate in the insight graph header which is something to keep in mind too if we plan to add more stuff there in the future.

> 6. I think it might be worth vertically centering the legend card if it's small. 

Done!

> 8. I found this bug when I created the chart in the screenshot where the legend shows empty. 

Fixed

 9. Think the legend should have a max width and then just allow scrolling. 

Max-height with overflow scrolling implemented.

alexkim205 committed:
make compare to legend more robust

alexkim205 committed:
fix test

alexkim205 committed:
add date inside of cell

clarkus commented:
> I've added a light grey bg to the insight graph container and standardized the bg colors we use across insights to bring it up to speed with our Figma mockups.

👍 

> I've changed it so that dates are now inside of each cell when compare functionality is on. Also fixed an existing bug where compare labels weren't being shown anywhere in our UX, which was making it difficult to distinguish current and previous rows. This let's me get rid of the compare column altogether. See screenshots above

While I see the value of days of the week, I don't think we need to show them directly at all. We use tooltips to translate other dates in the app, so we could just layer in the day of the week into that translation. 

> my 2c, zero-padding here will make the width ever so slightly wider than it has to be. We're already working with limited real estate in the insight graph header which is something to keep in mind too if we plan to add more stuff there in the future.

:+1:

paolodamico commented on frontend/src/lib/components/icons.tsx:
```diff
@@ -916,3 +916,14 @@ export function IconKeyboard(props: React.SVGProps<SVGSVGElement>): JSX.Element
     )
 }
+
+export function LegendIcon({ style }: { style?: CSSProperties }): JSX.Element {
```
Nit: `IconLegend` (for better consistency)

paolodamico committed:
adjust UI of table a bit

paolodamico commented on frontend/src/scenes/insights/InsightContainer.tsx:
```diff
@@ -175,17 +176,40 @@ export function InsightContainer(): JSX.Element {
+                        {activeView === InsightType.FUNNELS ||
+                            (activeView === InsightType.PATHS && (
+                                <Col>
+                                    <FunnelCanvasLabel />
```
We have a double check for `activeView` in the component, we should avoid this for easier maintenance

paolodamico commented on frontend/src/lib/components/InsightLegend/InsightLegend.tsx:
```diff
@@ -0,0 +1,69 @@
+    return (
+        <Button className="insight-legend-button" onClick={toggleInsightLegend}>
+            <LegendIcon />
+            <span className="insight-legend-button-title">{filters.legend_hidden ? 'Show' : 'Hide'} Legend</span>
```
```suggestion
            <span className="insight-legend-button-title">{filters.legend_hidden ? 'Show' : 'Hide'} legend</span>
```

paolodamico reviewed (approved):
It's a bit hard to fully evaluate such a large PR and make sure it doesn't introduce any new bugs, in general looks good to me! Some improvement suggestions below, but I'd suggest considering a separate PR,

- I adjusted the UI of the legend table with compare a bit, but I think we need @clarkus's eye here.
- Ideally we would sort the legend table with "Current" & "Previous" for each event/action in the same order (see different order here) for easy comparison.
    <img width="1166" alt="" src="https://user-images.githubusercontent.com/5864173/145079370-6cea2025-dc29-455a-9e61-6ab89a37e6ae.png">

paolodamico committed:
UI2

rcmarron commented on frontend/src/lib/components/DateFilter/DateFilter.tsx:
```diff
@@ -127,19 +136,30 @@ export function DateFilter({
 
                     return (
-                        <Select.Option key={key} value={key} label={makeLabel ? makeLabel(key) : undefined}>
+                        <Select.Option
```
I don't know the context for this change, but it feels weird to select an item in a drop down and then not have the drop down show the thing I clicked it. If we want to show the full date string, I wonder if it should be shown along with the range that the user clicked?

rcmarron commented on frontend/src/lib/components/InsightLegend/InsightLegend.tsx:
```diff
@@ -0,0 +1,69 @@
+                                <PHCheckbox
+                                    color={colorList[item.id]}
+                                    checked={visibilityMap[item.id]}
+                                    showIcon={indexedResults.length > 1}
```
Feels kind of like we're abusing the PHCheckbox by disbaling it and removing the checkbox. Should we just draw a box instead of adding the `showIcon` prop?

rcmarron replied:
On second thought - it doesn't feel too weird to just leave the checkbox functionality even if there is only one item. Maybe that's the right thing here?

alexkim205 replied:
The reason I changed it to hide the check was because I found myself clicking the checkbox trying to toggle it on and off. Maybe it makes sense not to disabled this on single series data altogether? It doesn't make sense that the user shouldn't have to power to hide everything if they wanted to?

rcmarron replied:
Yeah, not disabling on a single series makes sense to me.

rcmarron commented on frontend/src/scenes/insights/InsightTabs/InsightDisplayConfig.tsx:
```diff
@@ -92,60 +92,76 @@ export function InsightDisplayConfig({
                                     <CalendarOutlined /> {key}
                                 </>
                             )}
+                            isDateFormatted
```
Same thought on the date selector being odd here.

rcmarron commented on frontend/src/lib/components/InsightLegend/InsightLegend.tsx:
```diff
@@ -0,0 +1,69 @@
+    return (
+        <div className={clsx('insight-legend-menu', indexedResults.length <= 5 && 'short')}>
+            {indexedResults &&
+                indexedResults.map((item) => {
```
I'm not sure why this is happening, but the legend has a different order than the table below. You can see this if you turn on compare previous. The order here seems better, but either way they should probably be consistent.

alexkim205 replied:
I think I figured it out. The table comes presorted by total sum column and canceling this returns the table to it's original correct order. I've changed it so that we don't sort by total sum by default.

rcmarron reviewed (dismissed):
A lot of good fixes in here! A few thoughts in the comments.

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-tooltip-legend

alexkim205 commented on frontend/src/lib/components/CompareFilter/compareFilterLogic.ts:
```diff
@@ -26,6 +27,16 @@ export const compareFilterLogic = kea<compareFilterLogicType>({
+            if (compare !== undefined) {
+                actions.setCompare(compare)
+            }
+            if (insight === InsightType.LIFECYCLE || date_from === 'all') {
```
good call!

alexkim205 committed:
new feedback

alexkim205 committed:
remove presort table

clarkus reviewed (dismissed):
How is the size of the legend being determined? There's a good bit of extra space in this example and I don't see any content that would justify the space - just curious if we can optimize this more. Likewise I'm not sure if needs to be full height. I think it hsould size ot its contents. If those contents are taller than the available area, we can scroll the legend vertically. 

<img width="849" alt="Screen Shot 2021-12-08 at 11 37 49 AM" src="https://user-images.githubusercontent.com/254612/145272869-97562ad0-02e5-42d8-9dd0-5c8045a16c33.png">

This date formatting really competes with reading the metric values. I don't think we need to show the day of the week prominently here. Instead we should focus on showing the current time frame prominently with the comparison range in secondary position. For example a general pattern could be `Current (Comparison)` or **Dec 1** (Nov 23).

<img width="1544" alt="Screen Shot 2021-12-08 at 11 40 51 AM" src="https://user-images.githubusercontent.com/254612/145273276-1fb2c0e3-4e60-4b0e-be40-b905026afed5.png">

I think we should swap the legend values when breakdowns are applied. The table below shows each browser value primarily. The legend should match this. The most meaningful identifier is the value, not the event. 

<img width="756" alt="Screen Shot 2021-12-08 at 11 47 06 AM" src="https://user-images.githubusercontent.com/254612/145274448-88616997-a60d-4682-be4b-544e6618bbda.png">

EDsCODE commented:
RE: bug with breakdowns + compare. The link isn't showing me a graph that's broken or has compare set. I also just tried a breakdown with compare active and it seems to return as expected. Let me know what I should look at for the bug

alexkim205 commented:
> RE: bug with breakdowns + compare. The link isn't showing me a graph that's broken or has compare set. I also just tried a breakdown with compare active and it seems to return as expected. Let me know what I should look at for the bug

Huh it seems to be working now, looks like it was transient. Thanks for taking a look anyways!

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-tooltip-legend

alexkim205 committed:
address design feedback

alexkim205 commented:
Thanks @clarkus! Implemented all your feedback and this is what it looks like now:


<img width="1439" alt="Screen Shot 2021-12-08 at 4 42 59 PM" src="https://user-images.githubusercontent.com/13460330/145313930-05a49501-ed19-44c1-ac94-d3ee6e860694.png">

<img width="1429" alt="Screen Shot 2021-12-08 at 4 42 21 PM" src="https://user-images.githubusercontent.com/13460330/145313933-64d4a598-f081-43bd-bfbc-0a65fd8cce39.png">

<img width="1426" alt="Screen Shot 2021-12-08 at 4 42 14 PM" src="https://user-images.githubusercontent.com/13460330/145313934-ff085893-3b05-418b-a77f-dd3dd91451b7.png">

<img width="1426" alt="Screen Shot 2021-12-08 at 4 42 03 PM" src="https://user-images.githubusercontent.com/13460330/145313937-84190f92-2f35-4999-a921-20022626eb62.png">

alexkim205 committed:
fix default sorting

alexkim205 committed:
fix e2e tests
